### PR TITLE
[WIP] Refactor mvp

### PR DESCRIPTION
**Pull request by TheFynx**
Complete refactor of SLAPI for MVP.

* To Do: Create List of Changes and Updates

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -7,6 +7,7 @@ services:
     privileged: true
     volumes:
         - scripts:/usr/src/slapi/scripts
+        - brain:/usr/src/slapi/brain
```
Comment (aaronblythe) [inline docker-compose.yml:10]: Do these directories need to be created on the host before running?  If so should that be in the README for development?

File: lib/brain/redis.rb [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+      'HostConfig' => {
+        'PortBindings' => {
+          '6379/tcp' => [{ 'HostPort' => '6379', 'HostIp' => '0.0.0.0' }]
+        },
+        'Binds' => ["#{Dir.pwd}/brain/:/data"]
+      }
+    }
+    @container_hash['Entrypoint'] = @image.info['Config']['Entrypoint']
+    @container_hash['WorkingDir'] = @image.info['Config']['WorkingDir']
+    @container_hash['Labels'] = @image.info['Config']['Labels']
```
Comment (aaronblythe) [inline lib/brain/redis.rb]: Is there any thing in the @image.info['Config'] that should not be merged?

If not then use a simple Hash#merge. https://ruby-doc.org/core-2.4.0/Hash.html#method-i-merge.

You would have to play with this, but it would look something like:

    @container_hash.merge(@image.info['Config'])

File: lib/brain/redis.rb [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+class Brain
+  def initialize(settings)
+    @logger = Logger.new(STDOUT)
+    @logger.level = settings.logger_level
+  end
+
+  def brain
+    brain_check('slapi_brain')
+    @image = Docker::Image.create(fromImage: 'redis:3-alpine')
+    @container_hash = {
```
Comment (aaronblythe) [inline lib/brain/redis.rb]: Why the hash rockets here?

File: lib/brain/redis.rb [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+#     - URL For Redis determined by IP Comparison (Checking for Compose Environment)
+#  3. Enables Access to Brain
+#     - Query brain for specific key (i.e. Plugin User)
+#     - Query brain for specific hash (i.e. Plugin)
+#     - Delete Key from Brain (i.e. Plugin User)
+#     - Save Key/Value to Brain (i.e. Plugin User Bob )
+class Brain
+  def initialize(settings)
+    @logger = Logger.new(STDOUT)
+    @logger.level = settings.logger_level
```
Comment (aaronblythe) [inline lib/brain/redis.rb:25]: I believe we should log an issue to come back to for getting the logger set at a more global level.  Similar to some of the thoughts here: http://stackoverflow.com/questions/5995854/logging-in-sinatra

File: lib/plugins/plugin.rb [hunk 5]
```diff
@@ -18,34 +20,33 @@ class Plugin
...
+    @logger = Logger.new(STDOUT)
+    @logger.level = settings.logger_level
     @name = File.basename(file, '.*')
     @config = YAML.load_file(file)
-    @lang_settings = lang_settings
     @container = nil
-    @container_info = {}
+    @container_hash = {}
+    @container_hash['name'] = @name
+    @container_hash['HostConfig'] = {}
```
Comment (aaronblythe) [inline lib/plugins/plugin.rb]: Could just have:

    container_hash = {name: @name, HostConfig: {}}

File: lib/slapi.rb [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+  configure :production, :development, :test do
+    enable :logging
+  end
+
+  register Sinatra::SlapiRoutes::Plugin
+  register Sinatra::SlapiRoutes::Chat
+  register Sinatra::SlapiRoutes::Brain
+
+  @logger = Logger.new(STDOUT)
+  @logger.level = settings.logger_level
```
Comment (aaronblythe) [inline lib/slapi.rb:59]: The comment below will also play into setting up the logger in a better way.

File: lib/slapi.rb [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+  # Utilizes helper from brain/brain
+  @brain = Brain.new(settings)
+
+  # Load Plugins
+  # Utilizes Library from plugins/plugins
+  @plugins = Plugins.new(settings)
+
+  # Run Slapi Bot/Slack Client
+  # Utilizes Library from client/base
+  run
```
Comment (aaronblythe) [inline lib/slapi.rb]: Curious if the run occurs here if things are testable without being stuck in the async loop?  I honestly am not sure if it would be or not?

File: lib/slapi/client/base.rb [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    end
+    @client.start_async
+  end
+
+  def self.ping(data)
+    chat(
+      data,
+      title: 'Bot Check',
+      text: 'PONG',
+      color: '#229954'
```
Comment (aaronblythe) [inline lib/slapi/client/base.rb]: We should log an issue to follow up on the colors so they are named in some way an this implementation detail is not present throughout the code.

File: lib/slapi/client/base.rb [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+  def self.get_help(data)
+    # Remove when doing level 2 help or responding with specific plugin help
+    unless (data.text.include? "#{bot_prefix(data)}help ") || (@help_options['level'] == 2)
+      help_text = "Please use `@#{@bot_name} help plugin_name` for specific info"
+    end
+    if help_verify(data)
+      chat(
+        data,
+        pretext: help_text,
+        fallback: 'Your help has arrived!',
```
Comment (aaronblythe) [inline lib/slapi/client/base.rb]: An issue should be logged to externalize interaction strings like this for internationalization.  Whether we ever get to that is another story, however there are many interaction strings and we should note that as a known issue.

File: lib/slapi/helpers/base.rb [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+#     - Attachment Chat
+#  2. Set response channel based on where request type
+#     - Help request response in DM if set in Bot Config
+#     - No channel set for API requests (debug level only message)
+#     - Set to same channel as request
+#  3. Set Chat listener pre-fix base on channel
+#     - Listen for @bot, bot, or bot id as initiator if normal channel
+#     - Above plus any line that starts with a listener if direct message
+class Slapi
+  def self.chat(data = nil, channel = nil, attachment)
```
Comment (aaronblythe) [inline lib/slapi/helpers/base.rb]: For consistency attachment should be initialized also:

    self.chat(data = nil, channel = nil, attachment = [])

Otherwise the first two initializations are rather pointless.  More information here: http://stackoverflow.com/questions/9710303/rails-optional-argument

File: lib/slapi/helpers/base.rb [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+    @client.web_client.chat_postMessage(
+      channel: channel ? channel : channel_set(data),
+      as_user: true,
+      attachments: [attachment]
+    )
+  end
+
+  def self.chat_me(data = nil, channel = nil, attachment)
```
Comment (aaronblythe) [inline lib/slapi/helpers/base.rb]: Same initialization argument as above.

File: lib/slapi/helpers/base.rb [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+      channel: channel ? channel : channel_set(data),
+      attachments: [attachment]
+    )
+  end
+
+  def self.channel_set(data)
+    # Set channel to post based on dm_user option
+    if data.empty?
+      @logger.debug("Channel request wasn't given any data")
+      nil
```
Comment (aaronblythe) [inline lib/slapi/helpers/base.rb]: Is this even a valid case?  Should an exception be raised here?

File: lib/slapi/helpers/brain.rb [hunk 13]
```diff
@@ -0,0 +1,26 @@
+# frozen_string_literal: true
+
+# Slapi Brain Helper for Sinatra Extension Access
+class Slapi
+  def self.query_key(hash_name, key)
+    query = @brain.query_key(hash_name, key)
+    @logger.debug("Key retrieved for #{hash_name}")
```
Comment (aaronblythe) [inline lib/slapi/helpers/brain.rb]: If you are only logging hash_name, you can move this above the query line and not have to repeat.

If you want to log the query at debug level then that should be added.

File: lib/slapi/helpers/brain.rb [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+    query
+  end
+
+  def self.query_hash(hash_name)
+    query = @brain.query_hash(hash_name)
+    @logger.debug("Hash retrieved for #{hash_name}")
```
Comment (aaronblythe) [inline lib/slapi/helpers/brain.rb]: Same as above.

File: lib/slapi/helpers/plugins.rb [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+  end
+
+  def self.help_verify(data)
+    request = requested_plugin(data)
+    if request
+      @logger.debug("Slapi: Verifying requested plugin - #{request}")
+      @plugins.verify(request)
+    elsif !request
+      @logger.debug('Slapi: No specific plugin help requested')
+      true
```
Comment (aaronblythe) [inline lib/slapi/helpers/plugins.rb]: Just checking that this intention is to return true in this case?

File: lib/slapi/helpers/plugins.rb [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+    if data.text.include? ' '
+      requested_plugin = chat_data_sort(data)
+    elsif data.text.exclude? @client.self.id
+      @logger.debug('Slapi: Plugin called without args from DM')
+      requested_plugin = data.text
+    else
+      @logger.debug('Slapi: No matches found, no plugin requested')
+      requested_plugin = null
+    end
+    requested_plugin
```
Comment (aaronblythe) [inline lib/slapi/helpers/plugins.rb]: Given the current code this is extraneous.  However may be good to keep if you believe that "requested_plugin" will not be the final thing evaluated in all cases.

File: lib/slapi/helpers/plugins.rb [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+  end
+
+  def self.chat_data_sort(data)
+    # Create array based on spaces
+    data_array = data.text.split(' ')
+    if data.text.include? 'help'
+      @logger.debug('Slapi: Checking if help request includes a plugin')
```
Comment (aaronblythe) [inline lib/slapi/helpers/plugins.rb]: This seems like overkill.  Consider logging a ternary below:

    @logger.debug("Slapi: Help requested for #{requested_plugin ? requested_plugin : 'bogus plugin'}")

File: docker-compose.yml (hunk 1 above)
Comment (TheFynx) [reply docker-compose.yml:10]: No, everything is created by docker is needed. But this creating a shared volume, nothing on the host itself. I will have to add documentation to mount the `/usr/src/slapi/brain` locally if they wish to keep persistence outside of the compose setup.

File: lib/brain/redis.rb (hunk 2 above)
Comment (TheFynx) [reply lib/brain/redis.rb]: I'll have to dig into this, I honestly don't know. I specifically copied all the items that were needed. I just didn't think much about it after I got it working.

File: lib/brain/redis.rb (hunk 3 above)
Comment (TheFynx) [reply lib/brain/redis.rb]: Keeping with what used in the plugin.rb, but I refactored that to avoid overwrites I discovered while testing. I probably need to do that here too.

File: lib/plugins/plugin.rb (hunk 5 above)
Comment (TheFynx) [reply lib/plugins/plugin.rb]: 👍

File: lib/brain/redis.rb (hunk 4 above)
Comment (TheFynx) [reply lib/brain/redis.rb:25]: Yea, this was just a quick way to get the logger set the same across all classes.

File: lib/slapi.rb (hunk 6 above)
Comment (TheFynx) [reply lib/slapi.rb:59]: ^^ as mentioned before

File: lib/slapi.rb (hunk 7 above)
Comment (TheFynx) [reply lib/slapi.rb]: :man_shrugging:

File: lib/slapi/client/base.rb (hunk 8 above)
Comment (TheFynx) [reply lib/slapi/client/base.rb]: Yea, there should only be three colors. Would could possibly set constants for this as well.

File: lib/slapi/helpers/base.rb (hunk 12 above)
Comment (TheFynx) [reply lib/slapi/helpers/base.rb]: it has to be this way, I had to add this in. I can't be sure why at the moment, I'll have to go back through commit and runt he code without it again.

File: lib/slapi/helpers/brain.rb (hunk 13 above)
Comment (TheFynx) [reply lib/slapi/helpers/brain.rb]: Good call, probably just tired when I did this.

File: lib/slapi/helpers/plugins.rb (hunk 15 above)
Comment (TheFynx) [reply lib/slapi/helpers/plugins.rb]: Yup, intentional to make things trigger appropriately.

File: lib/slapi/helpers/plugins.rb (hunk 16 above)
Comment (TheFynx) [reply lib/slapi/helpers/plugins.rb]: I had to do this, beforehand it wasn't passing it correctly. I had to explicitly make it the last thing evaluated.

File: Dockerfile [hunk 18]
```diff
@@ -7,7 +7,7 @@ ENV APP_HOME /usr/src/slapi
 RUN mkdir -p $APP_HOME &&\
     mkdir -p $APP_HOME/log
 
-WORKDIR $APP_HOME
+WORKDIR /usr/src/slapi
```
Comment (aaronblythe) [inline Dockerfile:10]: Just curious on why this change?  Was the environment variable not working in this case?

File: README.md [hunk 19]
```diff
@@ -93,11 +108,18 @@ help:
 
 #### Local
 
-To run Sinatra locally simply run:
+To run Sinatra simply run:
+
+If using Bundler [Global](#global-install)
```
Comment (aaronblythe) [inline README.md]: Should both of these say "If using Bundler"?

File: Gemfile.lock [hunk 20]
```diff
@@ -144,19 +146,21 @@ GEM
...
-    slack-ruby-client (0.7.8)
+    slack-ruby-client (0.7.9)
       activesupport
       faraday (>= 0.9)
       faraday_middleware
       gli
       hashie
       json
       websocket-driver
+    sterile (1.0.10)
```
Comment (aaronblythe) [inline Gemfile.lock:157]: NOTE: could potentially use: http://api.rubyonrails.org/classes/ActiveSupport/Inflector.html

since it is already already pulled in by nature of using the slack-ruby-client: https://github.com/ImperialLabs/slapi/blob/master/Gemfile.lock#L148

File: lib/brain/redis.rb [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+    @container_hash = { name: 'slapi_brain', HostConfig: {} }
```
Comment (aaronblythe) [inline lib/brain/redis.rb]: Minor but appears @container_hash could just be a local variable in the `brain` method so it does not need to be carried around on the object.

File: lib/plugins/plugin.rb [hunk 22]
```diff
@@ -0,0 +1,201 @@
...
+    end
+
+    @logger.info("#{@name}: #{@help_hash ? 'Help list being built' : 'No help or labels found'}")
+
+    @help_hash&.each do |label, desc|
+      @help_list += '    ' + label + ' : ' + desc + "\n"
+    end
+  end
+
+  def help
```
Comment (aaronblythe) [inline lib/plugins/plugin.rb]: Should this just be an `attr_reader`?  Good explanation here: http://stackoverflow.com/questions/4370960/what-is-attr-accessor-in-ruby

File: lib/plugins/plugin.rb [hunk 23]
```diff
@@ -0,0 +1,201 @@
...
+        exec_passive(exec_data)
+      when 'active'
+        @container.exec([exec_data])
+      end
+    when 'api'
+      exec_api(data_from_chat, chat_text_array)
+    end
+  end
+
+  def exec_passive(exec_data)
```
Comment (aaronblythe) [inline lib/plugins/plugin.rb:150]: If the intention is that only exec should be called external to this object, then maybe `exec_passive` and `exec_api` should be private?

File: lib/plugins/plugin.rb [hunk 24]
```diff
@@ -0,0 +1,201 @@
...
+    HTTParty.post(@config['plugin']['api']['url'], body: payload, headers: @headers)
+    # else ?
+    # Error log and chat?
+    # Since it will only make it to this level if the bot was invoked
+    # then may it is appropriate to state that the bot does not understand?.
+  end
+
+  # Clears out existing container with the name planned to use
+  # Avoids this error:
+  # Uncaught exception: Conflict. The name "/hello_world" is already in use by container 23ee03db81c93cb7dd9eba206c3a7e.
```
Comment (aaronblythe) [inline lib/plugins/plugin.rb]: Have not got to tests yet, but curious is there is one that exercises this case yet.  If not then we should log an issue to get to eventually.

File: lib/plugins/plugins.rb [hunk 25]
```diff
@@ -0,0 +1,66 @@
...
+class Plugins
+  attr_reader :plugin_hash
+
+  def initialize(settings)
+    @plugin_hash = {}
+    @plugin_options = settings.plugins || {}
+    @help_options = settings.help || {}
+    @admin_options = settings.admin || {}
+    @bot_options = settings.bot || {}
+    @settings = settings
```
Comment (aaronblythe) [inline lib/plugins/plugins.rb]: Since settings is being saved as an instance variable, the 4 above are duplication (this making the object bigger).  Decide on whether want:
* to just use @settings.method in the code
* or to not store the full @settings object and use the named instance variables.

Either seems fine, delete the one not chosen.

File: lib/routes/brain.rb [hunk 26]
```diff
@@ -0,0 +1,81 @@
...
+        # @return [String] status or message "it worked"
+        slapi.post '/v1/save' do
+          raise 'missing plugin name' unless params[:plugin]
+          raise 'missing key' unless params[:key]
+          raise 'missing value' unless params[:value]
+
+          # Saves into brain as Plugin: plugin name (hash), key, value
+          slapi.save(params[:plugin], params[:key], params[:value])
+          status 200
+          { 'message' => 'it worked' }.to_json
```
Comment (aaronblythe) [inline lib/routes/brain.rb:30]: Not immediate, but at some point we should refactor this response message to be more meaningful and have error statuses if things are wrong.  This likely responds with 200 even if redis returns an error (https://redis.io/topics/protocol) from briefly looking at https://github.com/redis/redis-rb/blob/19bbbc8dc1b75cfbaf3dd7e0a92a9ff72f474fb5/lib/redis/client.rb#L121 I think that it raises an error that should be caught and then a status code and error message should be chosen based on that.  Currently in this the client of this call would not get information on why it failed and possibly would time out?

Comment (aaronblythe) [reply lib/routes/brain.rb:30]: Note: this is likely an issue that should be logged and not address as part of this pull request.

File: lib/routes/chat.rb [hunk 27]
```diff
@@ -0,0 +1,91 @@
...
+              fallback: params[:attachments][:fallback],
+              title: params[:attachments][:title],
+              title_link: params[:attachments][:title_link],
+              text: params[:attachments][:text],
+              color: params[:attachments][:color] ? params[:attachments][:color] : '#229954'
+            },
+            channel
+          )
+          status 200
+          { 'message' => 'it worked' }.to_json
```
Comment (aaronblythe) [inline lib/routes/chat.rb:46]: As a separate issue we should address passing back the error status of these to the client making the call as defined here: https://api.slack.com/methods/chat.postMessage

Appears that argument errors are thrown: https://github.com/slack-ruby/slack-ruby-client/blob/684533af2e43561323e55f14ce6ce0eef7b3fadb/lib/slack/web/api/endpoints/chat.rb#L72

And likely status codes pass through on errors.

**Review by aaronblythe — COMMENTED** [review]
Looks awesome!

File: Dockerfile (hunk 18 above)
Comment (TheFynx) [reply Dockerfile:10]: Coding Practices Apparently, installed a Dockerfile linter for grins. Small enough change.

File: Gemfile.lock (hunk 20 above)
Comment (TheFynx) [reply Gemfile.lock:157]: ^ Already made a comment in our coding room. Tried, did not work for smart qoutes.

File: lib/plugins/plugin.rb (hunk 24 above)
Comment (TheFynx) [reply lib/plugins/plugin.rb]: I don't think it does, it's does need to be added to test. But I was waiting to use codeclimate after the merge to create the issues (I like one button click solutions)

File: README.md (hunk 19 above)
Comment (TheFynx) [reply README.md]: :man_shrugging:
